<a href="https://colab.research.google.com/github/joiepark/assignment_esaa/blob/main/8%EC%9B%94%20%EC%8A%A4%ED%84%B0%EB%94%94%20-%20YB%20%EC%BB%A4%EB%A6%AC%ED%81%98%EB%9F%BC%20%EB%B3%B5%EC%8A%B5%202%ED%9A%8C%EC%B0%A8.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

### 회귀


*   지도학습은 분류와 회귀로 나뉜다. 이 두 가지 기법의 가장 큰 차이는 분류는 예측값이 카테고리와 같은 이산형 클래스 값이고 회귀는 연속형 숫자 값이라는 것이다.
*   W 파라미터의 개수가 적다면 고차원 방정식으로 비용 함수가 최소가 되는 W 변숫값을 도출할 수 있겠지만, W 파라미터가 많으면 고차원 방정식을 동원하더라도 해결하기가 어렵다. **경사 하강법**은 이러한 고차원 방정식에 대한 문제를 해결해주면서 비용 함수 RSS를 최소화하는 방법을 직관적으로 제공하는 방식이다. 점진적으로 반복적인 계산을 바탕으로 W 파라미터 값을 업데이트하며 오류 값이 최소가 되는 W 파라미터를 구하는 것이 경사 하강법의 메커니즘이다. 경사 하강법은 반복적으로 비용 함수의 반환 값, 즉 예측값과 실제 값의 차이가 작아지는 방향성을 가지고 W 파라미터를 지속해서 보정해나간다. 그리고 오류 값이 더 이상 작아지지 않으면 그 오류 값을 최소 비용으로 판단하고 그 때의 W 값을 최적 파라미터로 반환한다. 경사 하강법의 핵심은 "어떻게 하면 오류가 작아지는 방향으로 W 값을 보정할 수 있을까?"이다. 운동장에서 힘껏 야구공을 던지면 이 공의 속도가 처음에는 증가하다가 점차 감소하면서 땅에 떨어질 것이다. 처음에는 가속도가 계속 증가하면서 속도가 증가하고, 더 이상 가속도가 증가하지 않으면 그 때가 최고 속도이며, 그 후에는 가속도가 마이너스(-)가 되면서 속도가 떨어지고 마침내 공이 땅에 떨어진다. 가속도의 값은 속도의 미분이고, 속도와 같은 포물선 형태의 2차 함수의 최저점은 해당 2차 함수의 미분 값인 1차 함수의 기울기가 가장 최소일 때이다. 예를 들어 비용 함수가 아래로 볼록한 포물선 형태의 2차 함수라면, 경사 하강법은 최초 w에서부터 미분을 적용한 뒤 이 미분 값이 계속 감소하는 방향으로 순차적으로 w를 업데이트한다. 마침내 더 이상 미분된 1차 함수의 기울기가 감소하지 않는 지점을 비용함수가 최소인 지점으로 간주하고, 그 때의 w를 반환한다.
*   LinearRegression 클래스는 예측값과 실제 값의 RSS(Residual Sum of Squares)를 최소화해 OLS(Ordinary Least Squares) 추정 방식으로 구현한 클래스이다. LinearRegression 클래스는 fit() 메서드로 X, y 배열을 입력받으면 회귀 계수(Coefficients)인 W를 coef_ 속성에 저장한다.
*   입력 파라미터 (1) fit_intercept: 불린 값으로, 디폴트는 True이다. Intercept(절편) 값을 계산할 것인지 말지를 지정하며, False로 지정하면 intercpet가 사용되지 않고 0으로 지정된다. (2) normalize: 불린 값으로, 디폴트는 False이다. fit_intercept가 False일 경우 이 파라미터가 무시됨을 주의해야 한다. 만일 True이면 회귀를 수행하기 전에 입력 데이터 세트를 정규화한다.
*   속성 (1) coef_: fit() 메서드를 수행했을 때 회귀 계수가 배열 형태로 저장하는 속성. Shapes는 (Target 값 개수, 피처 개수)이다. (2) intercept__: intercept 값
Ordinary Least Squares 기반의 회귀 계수 계산은 입력 피처의 독립성에 많은 영향을 받는다. 피처 간의 상관관계가 매우 높은 경우 분산이 매우 커져서 오류에 매우 민감해진다. 이러한 현상을 다중 공선성 문제라고 한다. 일반적으로 상관관계가 높은 피처가 많은 경우 독립적인 중요한 피처만 남기고 제거하거나 규제를 적용한다. 또한 매우 많은 피처가 다중 공선성 문제를 가지고 있다면 PCA를 통해 차원 축소를 수행하는 것도 고려해볼 수 있다.
*   사이킷런은 RMSE를 제공하지 않으며, RMSE를 구하기 위해서는 MSE에 제곱근을 씌워 계산하는 함수를 직접 만들어야 한다.





*   비용 함수에 alpha 값으로 페널티를 부여해 회귀 계수 값의 크기를 감소시켜 과적합을 개선하는 방식을 규제(Regulation)라고 부른다. 규제는 크게 L2 방식과 L1 방식으로 구분된다. L2 규제는 W의 제곱에 대해 페널티를 부여하는 방식을 말하고, L2 규제를 적용한 회귀를 릿지(Ridge) 회귀라고 한다. 라쏘(Lasso 회귀)는 L1 규제를 적용한 회귀이다. L1 규제는 W의 절댓값에 대해 페널티를 부여한다. L1 규제를 적용하면 영향력이 크지 않은 회귀 계수 값을 0으로 변환한다.
*   로지스틱 회귀는 선형 회귀 방식을 분류에 적용한 알고리즘이다. 즉, 로지스틱 회귀는 분류에 사용된다. 로지스틱 회귀 역시 선형 계열이며, 로지스틱 회귀가 선형 회귀와 다른 점은 학습을 통해 선형 함수의 회귀 최적선을찾는 것이 아니라 시그모이드(Sigmoid) 함수 최적선을 찾고 이 시그모이드 함수의 반환 값을 확률로 간주해 확률에 따라 분류를 결정한다는 점이다.
*   사이킷런은 로지스틱 회귀를 위해 LogisticRegression 클래스를 제공한다. LogisticRegression 클래스의 회귀 계수 최적화는 경사 하강법 외에도 다양한 최적화 방안이 있다. LogisticRegression 클래스에서 solver 파라미터의 'lbfgs', 'liblinear', 'newton-cg', 'sag', 'saga' 값을 적용해 최적화를 선택할 수 있다. lbfgs: 사이킷런 버전 0.22부터 solver의 기본 설정값이다. 메모리 공간을 절약할 수 있고, CPU 코어 수가 많다면 최적화를 병렬로 수행할 수 있다. liblinear: 사이킷런 버전 0.21까지에서 solver의 기본 설정값이다. 다차원이고 작은 데이터 세트에서 효과적으로 동작하나, 국소 최적화(Local Minimum)에 이슈가 있으며 병렬로 최적화할 수 없다. newton-cg: 좀 더 정교한 최적화를 가능하게 하나, 대용량의 데이터에서 속도가 많이 느려진다. sag: Stochastic Average Gradient로, 경사 하강법 기반의 최적화를 적용한다. 대용량의 데이터에서 빠르게 최적화한다. saga: sag과 유사한 최적화 방식이며, L1 정규화를 가능케 한다. 이렇게 다양한 solver값들이 있지만 이들간의 성능 차이는 미비하며, 일반적으로 lbfgs 또는 liblinear를 선택하는 것이 대부분이다.



In [2]:
## 사이킷런의 LogisticRegression 클래스를 이용해 위스콘신 유방암 데이터 세트 기반에서 로지스틱 회귀로 암 여부를 판단한다
import pandas as pd
import matplotlib.pyplot as plt
%matplotlib inline

from sklearn.datasets import load_breast_cancer
from sklearn.linear_model import LogisticRegression

cancer = load_breast_cancer()  ## 전체 데이터를 생성한다



*   선형 회귀 계열의 로지스틱 회귀는 데이터의 정규 분포도에 따라 예측 성능 영향을 받을 수 있으므로, 데이터에 먼저 정규분포 형태의 표준 스케일링을 적용한 뒤, train_test_split()을 이용해 데이터 세트를 분리해야 한다.



In [3]:
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split

scaler = StandardScaler()
data_scaled = scaler.fit_transform(cancer.data)

X_train, X_test, y_train, y_test = train_test_split(data_scaled, cancer.target, test_size=0.3, random_state=0)

In [4]:
from sklearn.metrics import accuracy_score, roc_auc_score

## 로지스틱 회귀를 이용해 학습 및 예측 수행
## solver 인자값을 생성자로 입력하지 않으면 solver='lbfgs'
lr_clf = LogisticRegression()
lr_clf.fit(X_train, y_train)
lr_preds = lr_clf.predict(X_test)
lr_preds_proba = lr_clf.predict_proba(X_test)[:, 1]

## accuracy와 roc_auc를 측정한다
print('accuracy: {0:.3f}, roc_auc: {1:.3f}'.format(accuracy_score(y_test, lr_preds), roc_auc_score(y_test, lr_preds_proba)))

accuracy: 0.977, roc_auc: 0.995


solver가 lbfgs일 경우, 정확도가 0.977, ROC-AUC가 0.995로 도출되었다.

In [5]:
solvers = ['lbfgs', 'liblinear', 'newton-cg', 'sag', 'saga']

## 여러 개의 solver 값별로 LogisticRegression 학습 후 성능을 평가한다
for solver in solvers:
  lr_clf = LogisticRegression(solver = solver, max_iter=600)  ## 특정 solver는 최적화에 상대적으로 많은 반복 횟수가 필요할 수 있기 때문에 max_iter값을 600(큰 값)으로 설정한다
  lr_clf.fit(X_train, y_train)
  lr_preds = lr_clf.predict(X_test)
  lr_preds_proba = lr_clf.predict_proba(X_test)[:, 1]
  ## accuracy와 roc_auc를 측정한다
  print('solver:{0}, accuracy:{1:.3f}, roc_auc:{2:.3f}'.format(solver, accuracy_score(y_test, lr_preds), roc_auc_score(y_test, lr_preds_proba)))

solver:lbfgs, accuracy:0.977, roc_auc:0.995
solver:liblinear, accuracy:0.982, roc_auc:0.995
solver:newton-cg, accuracy:0.977, roc_auc:0.995
solver:sag, accuracy:0.982, roc_auc:0.995
solver:saga, accuracy:0.982, roc_auc:0.995


liblinear와 sag, saga일 때 정확도가 0.982, ROC-AUC가 0.995로 상대적인 성능 수치가 약간 높으나, 데이터 세트가 워낙 작기 때문에 개별 solver별 성능 결과의 차이는 크게 의미 있는 결과는 아니다.


*   solver와 max_iter 외에 사이킷런 LogisticRegression 클래스의 주요 하이퍼 파라미터로 penalty와 C가 있다. penalty는 규제(Regularization)의 유형을 설정하며 'l2'로 설정 시 L2 규제를, 'l1'으로 설정 시 L1 규제를 뜻한다. 기본은 'l2'이다. C는 규제 강도를 조절하는 alpha 값의 역수이다. 즉, C = 1/alpha이다. C 값이 작을수록 규제 강도가 크다.
*    L1, L2 규제의 경우 solver 설정에 따라 영향을 받는다. Liblinear, saga의 경우 L1, L2 규제가 모두 가능하지만, lbfgs, newton-cg, sag의 경우 L2 규제만 가능하다.



In [6]:
## GridSearchCV를 이용해 위스콘신 데이터 세트에서 solver, penalty, C를 최적화한다

from sklearn.model_selection import GridSearchCV

params = {'solver':['liblinear', 'lbfgs'], 'penalty':['l2', 'l1'], 'C':[0.01, 0.1, 1, 5, 10]}

lr_clf = LogisticRegression()

grid_clf = GridSearchCV(lr_clf, param_grid=params, scoring='accuracy', cv=3)
grid_clf.fit(data_scaled, cancer.target)
print('최적 하이퍼 파라미터:{0}, 최적 평균 정확도:{1:.3f}'.format(grid_clf.best_params_, grid_clf.best_score_))

최적 하이퍼 파라미터:{'C': 0.1, 'penalty': 'l2', 'solver': 'liblinear'}, 최적 평균 정확도:0.979


/usr/local/lib/python3.12/dist-packages/sklearn/model_selection/_validation.py:528: FitFailedWarning: 
15 fits failed out of a total of 60.
The score on these train-test partitions for these parameters will be set to nan.
If these failures are not expected, you can try to debug them by setting error_score='raise'.

Below are more details about the failures:
--------------------------------------------------------------------------------
15 fits failed with the following error:
Traceback (most recent call last):
  File "/usr/local/lib/python3.12/dist-packages/sklearn/model_selection/_validation.py", line 866, in _fit_and_score
    estimator.fit(X_train, y_train, **fit_params)
  File "/usr/local/lib/python3.12/dist-packages/sklearn/base.py", line 1389, in wrapper
    return fit_method(estimator, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/dist-packages/sklearn/linear_model/_logistic.py", line 1193, in fit
    solver = _check_solver

*   트리 기반의 회귀는 회귀 트리를 이용하는 것, 즉 회귀를 위한 트리를 생성하고 이를 기반으로 회귀 예측을 하는 것이다. 회귀 트리는 분류트리와 크게 다르지 않다. (* 결정 트리, 랜덤 포레스트, GBM, XGBoost, LightGBM 등의 모든 트리 기반의 알고리즘은 분류뿐만 아니라 회귀도 가능하다. 트리 생성이 CART(Classification And Regression Trees) 알고리즘에 기반하고 있기 때문이다.) 다만 리프 노드에서 예측 결정 값을 만드는 과정에 차이가 있는데, 분류 트리가 특정 클래스 레이블을 결정하는 것과는 달리 회귀 트리는 리프 노드에 속한 데이터 값의 평균값을 구해 회귀 예측값을 계산한다.

